asm/instruction.go

```diff
@@ -127,30 +131,53 @@ func (ins Instruction) Marshal(w io.Writer, bo binary.ByteOrder) (uint64, error)
 	return 2 * InstructionSize, nil
 }
 
-// RewriteMapPtr changes an instruction to use a new map fd.
+// AssociateMap associates a *ebpf.Map with this instruction.
 //
 // Returns an error if the instruction doesn't load a map.
-func (ins *Instruction) RewriteMapPtr(fd int) error {
-	if !ins.OpCode.IsDWordLoad() {
-		return fmt.Errorf("%s is not a 64 bit load", ins.OpCode)
+func (ins *Instruction) AssociateMap(m FDer) error {
+	if !ins.IsLoadFromMap() {
+		return errors.New("not a load from a map")
 	}
 
-	if ins.Src != PseudoMapFD && ins.Src != PseudoMapValue {
+	// Set fd to -1 since it is technically not yet rewritten
+	ins.encodeMapFD(-1)
```

Comment:
I'm not sure if we still need to do this. -1 flags to `fixupAndValidate` that a `Reference()` to a map hasn't been resolved into an fd yet, but as of this patch, that's no longer necessary.

Either:
- ins has a non-nil `Map()`, which can be resolved using `ins.Map().FD()` during marshaling
- the caller has set the FD manually (`> 0`)
- `ins.IsLoadFromMap()` is true and Constant is 0

I would suggest we drop this line and raise `ErrUnsatisfiedMapReference` when `Constant <= 0` in `fixupAndValidate()`.

Reply:
Good suggestion